snow/engine/common/timer.go

```diff
@@ -3,12 +3,113 @@
 
 package common
 
-import "time"
+import (
+	"sync"
+	"time"
+)
 
-// Timer describes the standard interface for specifying a timeout
-type Timer interface {
+// PreemptionSignal signals when to preempt the pendingTimeout of the timeout handler.
+type PreemptionSignal struct {
+	activateOnce sync.Once
+	initOnce     sync.Once
+	signal       chan struct{}
+}
+
+func (ps *PreemptionSignal) init() {
+	ps.signal = make(chan struct{})
+}
+
+// Listen returns a read-only channel that is closed when Preempt() is invoked.
+func (ps *PreemptionSignal) Listen() <-chan struct{} {
+	ps.initOnce.Do(ps.init)
+	return ps.signal
+}
+
+// Preempt causes any past and future calls of Listen to return a closed channel.
+func (ps *PreemptionSignal) Preempt() {
+	ps.initOnce.Do(ps.init)
+	ps.activateOnce.Do(func() {
+		close(ps.signal)
+	})
+}
+
+// timeoutScheduler schedules timeouts to be dispatched in the future.
+// Only a single timeout can be pending to be scheduled at any given time.
+// Once a preemption signal is closed, all timeouts are immediately dispatched.
+type timeoutScheduler struct {
+	newTimer         func(duration time.Duration) *time.Timer
+	onTimeout        func()
+	preemptionSignal <-chan struct{}
+	pendingTimeout   chan struct{}
+}
+
+// NewTimeoutScheduler constructs a new timeout scheduler with the given function to be invoked upon a timeout,
+// unless the preemptionSignal is closed and in which case it invokes the function immediately.
+func NewTimeoutScheduler(onTimeout func(), preemptionSignal <-chan struct{}, newTimer func(duration time.Duration) *time.Timer) *timeoutScheduler {
```

Comment:
q: since we only ever specify `newTimer` manually in tests in this package, would it make sense to remove this from the signature (initializing to `time.NewTimer` in this constructer) and then override the `newTimer` variable in the tests?

It seems like that would give a nicer UX for the prod code using this... But I don't have strong opinions here, just a small suggestion.

Reply:
Yes makes sense. That was also how it was in the previous incarnation of this code.